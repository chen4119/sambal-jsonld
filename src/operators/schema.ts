import { JsonLd, JSONLD_TYPE } from "../constants";
import { schemaTypeChildren } from "../schemaTypeChildren";

export function isSchemaType(jsonld: JsonLd, type: string, descendants = true) {
    if (!jsonld || !jsonld[JSONLD_TYPE] || !type) {
        return false;
    }
    const lowerJsonLdType = jsonld[JSONLD_TYPE].toLowerCase();
    const lowerType = type.toLowerCase();
    let children = [];
    if (descendants && schemaTypeChildren.has(lowerType)) {
        children = schemaTypeChildren.get(lowerType);
    }
    return lowerJsonLdType === lowerType || children.indexOf(lowerJsonLdType) >= 0;
}

export function getEntityByType(jsonLdList: JsonLd[], type: string, descendants = true) {
    if (Array.isArray(jsonLdList)) {
        for (const jsonld of jsonLdList) {
            if (isSchemaType(jsonld, type, descendants)) {
                return jsonld;
            }
        }
    }
    return null;
}

export function getEntitiesByType(jsonLdList: JsonLd[], type: string, descendants = true) {
    const entities = [];
    if (Array.isArray(jsonLdList)) {
        for (const jsonld of jsonLdList) {
            if (isSchemaType(jsonld, type, descendants)) {
                entities.push(jsonld);
            }
        }
    }
    return entities;
}