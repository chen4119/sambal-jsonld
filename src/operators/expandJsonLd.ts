import JsonLd from "../JsonLd";
import {
    JSONLD_ID
} from "../constants";

export function expandJsonLd(jsonld: unknown) {
    const jsonLd = new JsonLd(jsonld);
    return jsonLd.getNode(jsonld[JSONLD_ID], true, true);
}