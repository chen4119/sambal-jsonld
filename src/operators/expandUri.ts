import { JSONLD_CONTEXT, JSONLD_ID, JSONLD_TYPE } from "../constants";
import { isJsonLdRef, isObjectLiteral, normalizeUri, isAbsUri } from "../utils";

export function expandUri(jsonld: unknown, baseUrl?: string) {
    const derivedBase = (jsonld[JSONLD_CONTEXT] && jsonld[JSONLD_CONTEXT]["@base"]) ? 
        jsonld[JSONLD_CONTEXT]["@base"] : baseUrl;

    return derivedBase ? toAbsUri(derivedBase, jsonld) : jsonld;
}

function toAbsUri(baseUrl: string, jsonld: unknown) {
    if (Array.isArray(jsonld)) {
        return jsonld.map(j => toAbsUri(baseUrl, j));
    }

    if (isJsonLdRef(jsonld) && jsonld[JSONLD_ID] && !isAbsUri(jsonld[JSONLD_ID])) {
        jsonld[JSONLD_ID] = normalizeUri(`${baseUrl}${jsonld[JSONLD_ID]}`);
    } else if (isObjectLiteral(jsonld)) {
        for (const fieldName of Object.keys(jsonld)) {
            if (fieldName !== JSONLD_ID && fieldName !== JSONLD_TYPE && fieldName !== JSONLD_CONTEXT) {
                const fieldValue = jsonld[fieldName];
                jsonld[fieldName] = toAbsUri(baseUrl, fieldValue);
            }
        }
    }
    return jsonld;
}
