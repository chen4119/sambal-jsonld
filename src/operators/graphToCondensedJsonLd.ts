import JsonLd from "../JsonLd";
import {
    JSONLD_GRAPH
} from "../constants";
import {isObjectLiteral} from "../utils";

export function graphToCondensedJsonLd(jsonldGraph: unknown, id?: string) {
    if (isObjectLiteral(jsonldGraph) && jsonldGraph[JSONLD_GRAPH]) {
        const jsonld = new JsonLd(jsonldGraph);
        if (id) {
            return jsonld.getNode(id, true);
        }
        return jsonld.getGraphRootNodes();
    }
    return null;
}