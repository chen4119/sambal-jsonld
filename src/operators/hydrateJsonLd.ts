import {isObjectLiteral} from "../utils";
import JsonLd from "../JsonLd";

export async function hydrateJsonLd(json: unknown, fetcher: (url) => Promise<any>, context?: any) {
    if (Array.isArray(json)) {
        const resolvedArray = [];
        for (const item of json) {
            resolvedArray.push(await hydrateJsonLd(item, fetcher, context));
        }
        return resolvedArray;
    } else if (isObjectLiteral(json)) {
        const jsonld = new JsonLd(json, context);
        const hydratedJson = await jsonld.resolveJson(fetcher);
        if (hydratedJson) {
            for (const propName of Object.keys(hydratedJson)) {
                const propValue = await hydrateJsonLd(hydratedJson[propName], fetcher, context);
                hydratedJson[propName] = propValue;
            }
        }
        return hydratedJson;
    }
    return json;
}