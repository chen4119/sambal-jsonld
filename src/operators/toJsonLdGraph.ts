import JsonLd from "../JsonLd";
import {
    JSONLD_CONTEXT,
    JSONLD_GRAPH
} from "../constants";

export function toJsonLdGraph(jsonlds: unknown[], context: any) {
    if (jsonlds.length === 0) {
        return null;
    }
    const jsonLd = new JsonLd(jsonlds, context);
    const graph: Map<string, unknown> = jsonLd.flatten();
    return {
        [JSONLD_CONTEXT]: context,
        [JSONLD_GRAPH]: [...graph.values()]
    };
}