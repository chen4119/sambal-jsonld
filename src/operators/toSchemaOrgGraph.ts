import JsonLd from "../JsonLd";
import {
    SCHEMA_CONTEXT,
    JSONLD_CONTEXT,
    JSONLD_GRAPH
} from "../Constants";

export function toSchemaOrgGraph(schemaOrgJsonLds: object[]) {
    if (schemaOrgJsonLds.length === 0) {
        return null;
    }
    const jsonLd = new JsonLd(schemaOrgJsonLds, SCHEMA_CONTEXT);
    const graph: Map<string, object> = jsonLd.flatten();
    return {
        [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
        [JSONLD_GRAPH]: [...graph.values()]
    };
}