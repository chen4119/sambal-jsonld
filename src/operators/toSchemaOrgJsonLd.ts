import moment from 'moment';
import JsonLd, {Node} from "../JsonLd";
import {
    SCHEMA_CONTEXT,
    JSONLD_CONTEXT,
    JSONLD_TYPE,
    JSONLD_GRAPH,
    JSONLD_ID,
    SAMBAL_NAME,
    SCHEMA_PRIMITIVE_SET,
    SCHEMA_NUMBER,
    SCHEMA_INTEGER,
    SCHEMA_BOOL,
    SCHEMA_TEXT,
    SCHEMA_URL,
    SCHEMA_DATETIME,
    SCHEMA_DATE,
    SCHEMA_TIME
} from "../constants";
import {
    isSchemaOrgType,
    getSchemaOrgType,
    getSchemaOrgParentTypes,
    isUrl,
    makeAbsoluteIRI,
    isBlankNodeIRI
} from "../utils";

const DATE_FORMAT_REGEX = /^\d{4}-\d{1,2}-\d{1,2}$/;
const TIME_FORMAT_REGEX = /^\d{1,2}:\d{1,2}(:\d{1,2})?$/;

export function toSchemaOrgJsonLd(json: any, type: string, context?: any) {
    const typeId = `${SCHEMA_CONTEXT}/${type}`;
    if (!isSchemaOrgType(typeId)) {
        throw new Error(`schema.org ${type} not found`);
    }
    if (Array.isArray(json)) {
        const itemList = {
            [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
            [JSONLD_TYPE]: "ItemList",
            itemListElement: (json as any[]).map((item, index) => {
                const listItem = {
                    [JSONLD_TYPE]: "ListItem",
                    position: index + 1
                };
                if (typeof(item) === "string") {
                    listItem["url"] = item;
                } else {
                    listItem["item"] = objectToSchemaOrgType(item, type, context);
                }
                return listItem;
            })
        };
        return itemList;
    }
    return objectToSchemaOrgType(json, type, context);
}

function objectToSchemaOrgType(json: object, type: string, context?: any) {
    const jsonld = new JsonLd(json, context ? [SCHEMA_CONTEXT, context] : SCHEMA_CONTEXT);
    let nodeId;
    if (json[JSONLD_ID]) {
        nodeId = json[JSONLD_ID];
    } else {
        nodeId = '_:1';
    }
    if (jsonld.graphMap.has(nodeId)) {
        const node = jsonld.graphMap.get(nodeId);
        return {
            [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
            ...nodeToSchemaOrgType(node, type, jsonld.baseIRI)
        };
    }
    return null;
}

function getSchemaOrgProps(schema, typeProps) {
    for (const propName of Object.keys(schema)) {
        if (propName.startsWith("_")) {
            continue;
        }
        const propId = `${SCHEMA_CONTEXT}/${propName}`;
        typeProps[propId] = {
            name: propName,
            types: schema[propName].map(type => `${SCHEMA_CONTEXT}/${type}`)
        };
    }
}

function nodeToSchemaOrgType(node: Node, type: string, baseIRI: string) {
    const absTypeIRI = `${SCHEMA_CONTEXT}/${type}`;
    const typeSchema = getSchemaOrgType(absTypeIRI);
    const typeName = typeSchema[SAMBAL_NAME];
    const parents = getSchemaOrgParentTypes(absTypeIRI);
    const typeProps = {};
    getSchemaOrgProps(typeSchema, typeProps);
    for (const parentName of parents) {
        const parentSchema = getSchemaOrgType(`${SCHEMA_CONTEXT}/${parentName}`);
        getSchemaOrgProps(parentSchema, typeProps);
    }
    return populateSchemaOrgType(node, typeName, typeProps, baseIRI);
}

function populateSchemaOrgType(node: Node, typeName: string, typeProps: object, baseIRI: string) {
    const schemaOrgJsonLd = {
        [JSONLD_TYPE]: typeName
    };
    if (node.data[JSONLD_ID]) {
        schemaOrgJsonLd[JSONLD_ID] = getAbsoluteNodeID(baseIRI, node.data[JSONLD_ID]);
    }
    for (const link of node.links) {
        const prop = typeProps[link.edge];
        if (prop) {
            // TODO: check for node.id instead?
            const hasChildNode = link.node.links.length > 0;
            let propValue = null;
            if (hasChildNode) {
                const classTypeName = getClassTypeName(link.node.data, prop.types);
                // possible not to have a class type
                if (classTypeName) {
                    propValue = nodeToSchemaOrgType(link.node, classTypeName, baseIRI);
                }
            } else if (typeof(link.node.data) !== "undefined" && link.node.data !== null) {
                propValue = validatePrimitiveType(link.node.data, prop.types);
            }
            setSchemaOrgPropValue(schemaOrgJsonLd, prop.name, propValue, link.node.isArray);
        }
    }
    return schemaOrgJsonLd;
}

function getAbsoluteNodeID(baseIRI: string, nodeId: string) {
    if (isUrl(nodeId)) {
        return nodeId;
    }
    if (baseIRI && !isBlankNodeIRI(nodeId)) {
        return makeAbsoluteIRI(baseIRI, nodeId);
    }
    return nodeId;
}

function setSchemaOrgPropValue(schemaOrgJsonLd: object, propName: string, propValue: any, isArray: boolean) {
    if (propValue) {
        if (isArray) {
            schemaOrgJsonLd[propName] = Array.isArray(schemaOrgJsonLd[propName]) ? [...schemaOrgJsonLd[propName], propValue] : [propValue];
        } else {
            schemaOrgJsonLd[propName] = propValue;
        }
    }
}

function validatePrimitiveType(value: any,  propTypes: string[]) {
    const valueType = getPrimitiveType(value);
    if (!valueType) {
        console.error(`Unknown primitive type for value: ${value}`);
        return null;
    }
    switch (valueType) {
        case SCHEMA_DATETIME:
            return validateDate(value, propTypes);
        case SCHEMA_NUMBER:
            return validateNumber(value, propTypes);
        case SCHEMA_TEXT:
            return validateString(value, propTypes);
        default:
            return propTypes.indexOf(valueType) >= 0 ? value : null;
    }
}

function validateDate(value: Date, propTypes: string[]) {
    const isAllowDateTime = propTypes.indexOf(SCHEMA_DATETIME) >= 0;
    const isAllowDate = propTypes.indexOf(SCHEMA_DATE) >= 0;
    const isAllowTime = propTypes.indexOf(SCHEMA_TIME) >= 0;
    if (isAllowDateTime) {
        return moment(value).toISOString();
    } else if (isAllowDate) {
        return moment(value).format("YYYY-MM-DD");
    } else if (isAllowTime) {
        return moment(value).format("HH:mm:ss");
    }
    return null;
}

function validateNumber(value: number, propTypes: string[]) {
    const isAllowNumber = propTypes.indexOf(SCHEMA_NUMBER) >= 0;
    const isAllowInteger = propTypes.indexOf(SCHEMA_INTEGER) >= 0;
    if (isAllowNumber || (isAllowInteger && Number.isInteger(value))) {
        return value;
    }
    return null;
}

function validateString(value: string, propTypes: string[]) {
    const isAllowDateTime = propTypes.indexOf(SCHEMA_DATETIME) >= 0;
    const isAllowDate = propTypes.indexOf(SCHEMA_DATE) >= 0;
    const isAllowTime = propTypes.indexOf(SCHEMA_TIME) >= 0;
    if (isAllowDateTime || isAllowDate || isAllowTime) {
        const d = moment(value, moment.ISO_8601);
        if (!d.isValid()) {
            return null;
        }
        if ((isAllowDate && DATE_FORMAT_REGEX.test(value)) || (isAllowTime && TIME_FORMAT_REGEX.test(value))) {
            return value;
        }  
        return validateDate(d.toDate(), propTypes);
    }
    return value;
}

function getPrimitiveType(value: any) {
    const typeOf = typeof(value);
    if (moment.isDate(value)) {
        return SCHEMA_DATETIME;
    } else if (typeOf === "number") {
        return SCHEMA_NUMBER;
    } else if (typeOf === "boolean") {
        return SCHEMA_BOOL;
    } else if (typeOf === "string") {
        if (isUrl(value)) {
            return SCHEMA_URL;
        }
        return SCHEMA_TEXT;
    }
    return null; // should not happen
}

// potential has more than one class type, just get first one
function getClassTypeName(json: any, propTypes: string[]) {
    if (json[JSONLD_TYPE]) {
        const typeId = `${SCHEMA_CONTEXT}/${json[JSONLD_TYPE]}`;
        for (const validTypeId of propTypes) {
            const isDescendant = isDescendantOf(typeId, validTypeId);
            if (isDescendant) {
                return json[JSONLD_TYPE];
            }
        }
    }
    return null;
}

function isDescendantOf(absSchemaIRI: string, absParentIRI: string) {
    if (absSchemaIRI.toLowerCase() === absParentIRI.toLowerCase()) {
        return true;
    }
    if (isSchemaOrgType(absSchemaIRI) && isSchemaOrgType(absParentIRI)) {
        const schema = getSchemaOrgType(absSchemaIRI);
        const parents = getSchemaOrgParentTypes(absSchemaIRI);
        for (const parentName of parents) {
            if (`${SCHEMA_CONTEXT}/${parentName}` === absParentIRI) {
                return true;
            }
        }
    }
    return false;
}

