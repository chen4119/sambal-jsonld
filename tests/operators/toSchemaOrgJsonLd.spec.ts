import {toSchemaOrgJsonLd} from "../../src/operators/toSchemaOrgJsonLd";

describe('toSchemaOrgJsonLd', () => {
    
    describe('person', () => {
        const person = {
            "name": "Jane Doe",
            "firstName": "Wan Chun",
            "telephone": "(425) 123-4567",
            "url": "http://www.janedoe.com",
            "sameAs": [
                "https://github.com/chen4119",
                "https://www.linkedin.com/in/wan-chun-chen-9a95a010"
            ]
        };
    
        it("to schema.org", () => {
            const schema = toSchemaOrgJsonLd(person, "Person");
            expect(schema).toMatchSnapshot();
        });

        it("to schema.org with givenName", () => {
            const schema = toSchemaOrgJsonLd(person, "Person", {firstName: "http://schema.org/givenName"});
            expect(schema).toMatchSnapshot();
        });
    });

    describe('blogpost', () => {
        const blogPost = {
            "@id": "blog1",
            headline: "Fourth post",
            description: "Some description",
            author: {
                "@id": "fasdf",
                "@type": "Person",
                "name": "Jane Doe",
                "telephone": "(425) 123-4567",
                "url": "http://www.janedoe.com"
            },
            dateCreated: "2019-09-06",
            keywords: ["schema.org"],
            version: 3
        };
    
        it("to schema.org", () => {
            const schema = toSchemaOrgJsonLd(blogPost, "BlogPosting");
            expect(schema).toMatchSnapshot();
        });

        it("to schema.org with absolute @id URI", () => {
            const schema = toSchemaOrgJsonLd(blogPost, "BlogPosting", {"@base": "http://chen4119.me#"});
            expect(schema).toMatchSnapshot();
        });

        it("remove dateCreated if format is invalid", () => {
            const customBlogPost = {
                ...blogPost,
                dateCreated: "abs"
            }
            const schema = toSchemaOrgJsonLd(customBlogPost, "BlogPosting");
            expect(schema.dateCreated).not.toBeDefined();
        });

        it("can use datetime format for dateCreated", () => {
            const customBlogPost = {
                ...blogPost,
                dateCreated: "2019-09-06T18:39:29"
            }
            const schema = toSchemaOrgJsonLd(customBlogPost, "BlogPosting");
            expect(schema.dateCreated).toBeDefined();
        });

        it("can use timezone datetime format for dateCreated", () => {
            const customBlogPost = {
                ...blogPost,
                dateCreated: "2019-09-06T18:39:29-05:00"
            }
            const schema = toSchemaOrgJsonLd(customBlogPost, "BlogPosting");
            expect(schema.dateCreated).toBeDefined();
        });

        it("can use Date for dateCreated", () => {
            const customBlogPost = {
                ...blogPost,
                dateCreated: new Date()
            }
            const schema = toSchemaOrgJsonLd(customBlogPost, "BlogPosting");
            expect(schema.dateCreated).toBeDefined();
        });
    });

});
