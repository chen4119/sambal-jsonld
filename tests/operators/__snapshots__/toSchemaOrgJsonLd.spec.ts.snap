// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toSchemaOrgJsonLd blogpost to schema.org 1`] = `
Object {
  "@context": "https://schema.org",
  "@id": "blog1",
  "@type": "BlogPosting",
  "author": Object {
    "@id": "fasdf",
    "@type": "Person",
    "name": "Jane Doe",
    "telephone": "(425) 123-4567",
    "url": "http://www.janedoe.com",
  },
  "dateCreated": "2019-09-06",
  "description": "Some description",
  "headline": "Fourth post",
  "keywords": Array [
    "schema.org",
  ],
  "version": 3,
}
`;

exports[`toSchemaOrgJsonLd blogpost to schema.org with absolute @id URI 1`] = `
Object {
  "@context": "https://schema.org",
  "@id": "http://chen4119.me#blog1",
  "@type": "BlogPosting",
  "author": Object {
    "@id": "http://chen4119.me#fasdf",
    "@type": "Person",
    "name": "Jane Doe",
    "telephone": "(425) 123-4567",
    "url": "http://www.janedoe.com",
  },
  "dateCreated": "2019-09-06",
  "description": "Some description",
  "headline": "Fourth post",
  "keywords": Array [
    "schema.org",
  ],
  "version": 3,
}
`;

exports[`toSchemaOrgJsonLd person to schema.org 1`] = `
Object {
  "@context": "https://schema.org",
  "@type": "Person",
  "name": "Jane Doe",
  "sameAs": Array [
    "https://github.com/chen4119",
    "https://www.linkedin.com/in/wan-chun-chen-9a95a010",
  ],
  "telephone": "(425) 123-4567",
  "url": "http://www.janedoe.com",
}
`;

exports[`toSchemaOrgJsonLd person to schema.org with givenName 1`] = `
Object {
  "@context": "https://schema.org",
  "@type": "Person",
  "name": "Jane Doe",
  "sameAs": Array [
    "https://github.com/chen4119",
    "https://www.linkedin.com/in/wan-chun-chen-9a95a010",
  ],
  "telephone": "(425) 123-4567",
  "url": "http://www.janedoe.com",
}
`;
